components/autocomplete/index.js

```diff
@@ -34,86 +99,179 @@ class Autocomplete extends Component {
 		this.bindNode = this.bindNode.bind( this );
 		this.select = this.select.bind( this );
 		this.reset = this.reset.bind( this );
-		this.onBlur = this.onBlur.bind( this );
 		this.search = this.search.bind( this );
 		this.setSelectedIndex = this.setSelectedIndex.bind( this );
 
 		this.state = this.constructor.getInitialState();
+		this.allOptions = {};
 	}
 
 	bindNode( node ) {
 		this.node = node;
 	}
 
 	select( option ) {
-		const { onSelect } = this.props;
+		const { open, range, lookup } = this.state;
+		const { onSelect } = open;
 
 		this.reset();
 
 		if ( onSelect ) {
-			onSelect( option );
+			onSelect( option.value, range, lookup );
 		}
 	}
 
 	reset() {
 		this.setState( this.constructor.getInitialState() );
 	}
 
-	onBlur( event ) {
-		// Check if related target is not within, in the case that the user is
-		// selecting an option by button click
-		if ( ! this.node.contains( event.relatedTarget ) ) {
-			this.reset();
+	// this method is separate so it can be overrided in tests
+	getCursor( container ) {
+		const selection = window.getSelection();
+		if ( selection.isCollapsed && container.contains( selection.anchorNode ) ) {
+			return {
+				node: selection.anchorNode,
+				offset: selection.anchorOffset,
+			};
 		}
+		return null;
 	}
 
-	search( event ) {
-		const { triggerPrefix } = this.props;
-		const { isOpen } = this.state;
-
-		const value = event.target.textContent;
-
-		// Currently search supports only if input's entire value is match
-		let pattern = '^';
-
-		if ( triggerPrefix ) {
-			// Trigger prefix may contain reserved characters of a regular
-			// expression, so escape
-			pattern += escapeStringRegexp( triggerPrefix );
-		}
-
-		// Create matching group for the search value
-		pattern += '([^\\s]*)$';
-
-		pattern = new RegExp( pattern );
-
-		let match, search;
-		if ( ( match = value.match( pattern ) ) ) {
-			search = new RegExp( escapeStringRegexp( match[ 1 ] ), 'i' );
-		}
+	// this method is separate so it can be overrided in tests
+	createRange( startNode, startOffset, endNode, endOffset ) {
+		const range = document.createRange();
+		range.setStart( startNode, startOffset );
+		range.setEnd( endNode, endOffset );
+		return range;
+	}
 
-		if ( isOpen ) {
+	search( event ) {
+		const createRange = this.createRange;
+		const { completers: allCompleters } = this.props;
+		const allowAnything = () => true;
+		const { open } = this.state;
+		// ensure that the cursor location is unambiguous
+		const container = event.target;
+		const cursor = this.getCursor( container );
+		// look for the trigger prefix and search query just before the cursor location
+		const match = ( function() {
```

Comment:
Note:
match() does the following:
1. Search backwards from the cursor looking for the first space or element boundary (ie text highlighted bold)
2. Then matches the trigger prefix of the completers  to the text immediately after the space/element boundary.
In the process it will filter the list of completers by 2 other critera:
1. All text nodes included in the match must be considered acceptable by the completer - this helps us check that we're not trying to create a link inside another link.
2. The range before and after the match must be acceptable to the completer - this helps us ensure that we're not going to replace a block that actually has content in it.

Reply:
Should this just be a separate function? `getMatch` ? The IIFE might not be an obvious pattern for the uninitiated reader.

Reply:
I extracted it out as findMatch